core/src/main/java/org/javarosa/form/api/FormEntrySessionReplayer.java

```diff
@@ -26,47 +27,58 @@ public static void tryReplayingFormEntry(FormEntryController formEntryController
                                              FormEntrySession formEntrySession) {
         FormEntrySessionReplayer replayer =
                 new FormEntrySessionReplayer(formEntryController, formEntrySession);
-        if (replayer.isRestoringFormSession()) {
+        if (replayer.hasSessionToReplay()) {
             replayer.replayForm();
         }
     }
 
-    private boolean isRestoringFormSession() {
+    private boolean hasSessionToReplay() {
         return formEntrySession != null && formEntrySession.size() > 0;
     }
 
+    /**
+     * TODO AMS: If the question corresponding to the stopping ref has been removed, this will
+     * never return true and replay will take the user all the way to the end of the form
+     */
+    private boolean reachedEndOfReplay(String lastQuestionRefReplayed) {
+        return lastQuestionRefReplayed.equals(formEntrySession.getStopRef());
+    }
```

Comment:
Doesn't that essentially come down to the same thing as checking if the vector of actions is empty?  If one of the questions that was originally answered in the form entry session has been subsequently deleted, the count would never reach zero. This is at least a little more robust because it only fails if 1 specific question was deleted.